src/named_pipe.rs

```diff
@@ -0,0 +1,208 @@
+
+        Ok(NamedPipe {
+            component_input_fifo_path,
+            component_output_fifo_path,
+
+            component_input_fifo_file: None,
+            component_output_fifo_file: None,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct NamedPipe {
+    component_input_fifo_path: PathBuf,
+    component_output_fifo_path: PathBuf,
+
+    component_input_fifo_file: Option<File>,
+    component_output_fifo_file: Option<File>,
+}
+
+const PIPE_CREATION_TIMEOUT_MS: u64 = 100;
+const PIPE_IO_TIMEOUT_MS: u64 = 1000;
+const PIPE_POLL_INTERVAL_MS: u64 = 2;
+
+const BUF_SIZE: usize = 256;
+
+impl NamedPipe {
+    fn get_fds(&mut self) -> Result<(RawFd, RawFd), WorkerError> {
+        let deadline = Instant::now() + Duration::from_millis(PIPE_CREATION_TIMEOUT_MS);
+
+        while self.component_output_fifo_file.is_none() && Instant::now() < deadline {
+            let c_out_res = OpenOptions::new()
+                .read(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_output_fifo_path);
+
+            debug!("Opening component output {:?}", c_out_res);
+
+            self.component_output_fifo_file = c_out_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
+        }
+
+        while self.component_input_fifo_file.is_none() && Instant::now() < deadline {
+            let c_in_res = OpenOptions::new()
+                .write(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_input_fifo_path);
+
+            debug!("Opening component input {:?}", c_in_res);
+
+            self.component_input_fifo_file = c_in_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
+        }
+
+        if let (Some(c_in), Some(c_out)) =
+            (&self.component_input_fifo_file, &self.component_output_fifo_file)
+        {
+            Ok((c_in.as_raw_fd(), c_out.as_raw_fd()))
+        } else {
+            Err(WorkerErrorKind::OperationTimedOut.into())
+        }
+    }
+
+    pub fn write(&mut self, v: &[u8]) -> Result<(), WorkerError> {
+        let (c_in_fd, _) = self.get_fds()?;
+
+        let deadline = Instant::now() + Duration::from_millis(PIPE_IO_TIMEOUT_MS);
+
+        assert_eq!(v.last(), Some(&b'\n'));
+
+        let mut write_idx = 0;
+        while write_idx < v.len() && Instant::now() < deadline {
+            // Wait until ready
+            let poll_flags = PollFlags::POLLOUT;
+            poll(
+                &mut [PollFd::new(c_in_fd, poll_flags)],
+                (deadline - Instant::now()).as_millis() as c_int,
+            )?;
+
+            // Then write the bytes
+            let written_bytes = write(c_in_fd, &v[write_idx..])?;
+            write_idx += written_bytes;
+        }
+
+        // If we didn't write everything, we timed out
+        if write_idx < v.len() {
+            return Err(WorkerErrorKind::OperationTimedOut.into());
+        }
+
+        Ok(())
+    }
```

Comment:
Is this because there's a possible timeout?

Reply:
Yep, that's why we need an error :)